cmd/root_test.go

```diff
@@ -1531,6 +1531,109 @@ func TestGit(t *testing.T) {
 	)
 }
 
+func TestTreeRootExclusivity(t *testing.T) {
+	tempDir := test.TempExamples(t)
+	configPath := filepath.Join(tempDir, "/treefmt.toml")
+
+	formatterConfigs := map[string]*config.Formatter{
+		"echo": {
+			Command:  "echo", // will not generate any underlying changes in the file
+			Includes: []string{"*"},
+		},
+	}
+
+	test.ChangeWorkDir(t, tempDir)
+
+	assertExclusiveFlag := func(as *require.Assertions, err error) {
+		as.ErrorContains(err,
+			"if any flags in the group [tree-root tree-root-cmd tree-root-file] are set none of the others can be;",
+		)
+	}
+
+	assertExclusiveConfig := func(as *require.Assertions, err error) {
+		as.ErrorContains(err,
+			"only one of tree-root, tree-root-cmd or tree-root-file can be specified",
+		)
+	}
+
+	envValues := map[string][]string{
+		"tree-root":      {"TREEFMT_TREE_ROOT", "bar"},
+		"tree-root-cmd":  {"TREEFMT_TREE_ROOT_CMD", "'echo buzz'"},
+		"tree-root-file": {"TREEFMT_TREE_ROOT_FILE", ".git/config"},
+	}
+
+	flagValues := map[string][]string{
+		"tree-root":      {"--tree-root", "bar"},
+		"tree-root-cmd":  {"--tree-root-cmd", "'echo buzz'"},
+		"tree-root-file": {"--tree-root-file", ".git/config"},
+	}
+
+	configValues := map[string]func(*config.Config){
+		"tree-root": func(cfg *config.Config) {
+			cfg.TreeRoot = "bar"
+		},
+		"tree-root-cmd": func(cfg *config.Config) {
+			cfg.TreeRootCmd = "'echo buzz'"
+		},
+		"tree-root-file": func(cfg *config.Config) {
+			cfg.TreeRootFile = ".git/config"
+		},
+	}
+
+	permutations := [][]string{
```

Comment:
nit: "permutations" is not the right word here. "permutation" is about the order of things, and this is not about ordering.

Mathematically, I think it's all subsets of the 3 parameters with more than 1 element (`all_subsets - subsets_of_length_1 - subsets_of_length_0 = 2 ** 3 - 3 - 1 = 4`). That's a mouthful, though. Perhaps "conflictingArgSets"?

I'd also be fine with not actually checking all conflicts. This list would start to get pretty silly if we added a 4th parameter, for instance.

Reply:
I've gone with `invalidCombinations`.